lib/Optimizer/Transforms/PhaseFolding.cpp

```diff
@@ -0,0 +1,656 @@
+        return a->isBeforeInBlock(b);
+      };
+      std::sort(ordered_ops.begin(), ordered_ops.end(), less);
+    }
+
+    return ordered_ops;
+  }
+
+  /// @brief Gets the number of RZs in the subcircuit
+  size_t getNumRotations() { return num_rot_gates; }
+
+  /// @returns The percentage of operations in the subcircuit
+  /// that are `quake.rz`s.
+  float getRotationWeight() {
+    return (float)getNumRotations() / (float)getNumOps();
+  }
+
+  /// @brief Gets the number of operations in the subcircuit
+  size_t getNumOps() { return ops.size(); }
+};
+
+struct PhaseVariable {
+public:
+  size_t idx;
+  // Q: do we really need the initial_wire here? I think it's just useful for
+  // debugging
+  Value initial_wire;
+  PhaseVariable(size_t index, Value wire) : idx(index), initial_wire(wire) {}
+
+  bool operator==(PhaseVariable other) { return idx == other.idx; }
+};
+
+/// A `Phase` is an exclusive sum of all of the `PhaseVariable`s involved in the
+/// current state of a qubit, as well as 1, representing inversion from a Not
+/// gate. The simplest Phase contains exactly the `PhaseVariable` representing
+/// the initial state of a qubit in a subcircuit. There are two operations on
+/// `Phase`s to generate new `Phase`s: `Phase::sum` sums two Phases,
+/// corresponding to the effect of a CNot on the target qubit. `Phase::invert`
+/// inverts a Phase, corresponding to the effect of a Not a qubit.
+///
+/// Generally, a Phase is an exclusive sum of products.
+/// However, our Phases are currently only exclusive sums;
+/// products are not currently supported.
+///
+/// Any two rotations on qubits with equal Phases can be merged into one
+/// rotation.
+class Phase {
+  SetVector<PhaseVariable *> vars;
+  bool isInverted;
+
+public:
+  Phase() : isInverted(false) {}
+
+  Phase(PhaseVariable *var) : isInverted(false) { vars.insert(var); }
+
+  /// @brief Two phases are equal if they contain exactly the same vars
+  /// and have the same inversion flag.
+  bool operator==(Phase other) {
+    for (auto var : vars)
+      if (!other.vars.contains(var))
+        return false;
+    for (auto var : other.vars)
+      if (!vars.contains(var))
+        return false;
+    return isInverted == other.isInverted;
+  }
+
+  /// @brief Returns a new phase equal to the sum of `p1` and `p2`
+  /// @returns A new phase containing the exclusive or of `p1` and `p2`
+  static Phase sum(Phase &p1, Phase &p2) {
+    Phase new_phase = Phase();
+    for (auto var : p1.vars)
+      new_phase.vars.insert(var);
+    for (auto var : p2.vars)
+      if (new_phase.vars.contains(var))
+        new_phase.vars.remove(var);
+      else
+        new_phase.vars.insert(var);
+    new_phase.isInverted = (p1.isInverted != p2.isInverted);
+    return new_phase;
+  }
+
+  /// @brief Returns a new phase equal to `p1` with the opposite inversion flag
+  static Phase invert(Phase &p1) {
+    auto new_phase = Phase();
+    for (auto var : p1.vars)
+      new_phase.vars.insert(var);
```

Comment:
Q: Could we write this as:

```suggestion
    new_phase.vars.insert(p1.vars.begin(), p1.vars.end());
```
?